File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/component/ProgressBar.java
package io.github.biligoldenwater.midiplayer.utils.particlegui.component;

import io.github.biligoldenwater.midiplayer.utils.particlegui.PixelColor;
import io.github.biligoldenwater.midiplayer.utils.particlegui.interfaces.Component;
import org.bukkit.Color;

public class ProgressBar extends Component {
    private boolean fitParent = false;
    private boolean isHorizontal;
    private double percent;
    private PixelColor color;

    public ProgressBar(int x, int y, int width, int height, boolean isHorizontal) {
        super(x, y, width, height, false);

        this.color = new PixelColor(Color.BLUE);
        this.isHorizontal = isHorizontal;

        render();
    }

    public ProgressBar(int x, int y, int width, int height) {
        this(x, y, width, height, true);
    }

    @Override
    public void render() {
        if (fitParent && getParent() != null) {
            setSize(getParent().getClientWidth(), getParent().getClientHeight());
        } else {
            this.pixels = new PixelColor[getWidth() * getHeight()];
        }
        if (!isHorizontal) {
            for (int i = 0; i < pixels.length * percent * 0.01; i++) {
                pixels[i] = color;
            }
        } else {
            long pixelNum = 1;

            int width = getWidth();
            int height = getHeight();

            for (int x = 0; x < width; x++) {
                if (pixelNum > pixels.length * percent * 0.01) break;
                for (int y = height - 1; y >= 0; y--) {
                    if (pixelNum > pixels.length * percent * 0.01) break;
                    pixels[x + y * width] = color;
                    pixelNum++;
                }
            }
        }
    }

    public double getPercent() {
        return percent;
    }

    public void setPercent(double percent) {
        onChange();
        this.percent = Math.min(percent, 100f);
    }

    public PixelColor getColor() {
        return color;
    }

    public void setColor(PixelColor color) {
        onChange();
        this.color = color;
    }

    public boolean isHorizontal() {
        return isHorizontal;
    }

    public void setHorizontal(boolean horizontal) {
        onChange();
        isHorizontal = horizontal;
    }

    public boolean isFitParent() {
        return fitParent;
    }

    public void setFitParent(boolean fitParent) {
        this.fitParent = fitParent;
    }
}

File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/component/ProgressBar.kt
package io.github.biligoldenwater.midiplayer.utils.particlegui.component

import io.github.biligoldenwater.midiplayer.utils.particlegui.PixelColor
import io.github.biligoldenwater.midiplayer.utils.particlegui.interfaces.Component
import org.bukkit.Color

class ProgressBar constructor(x: Int, y: Int, width: Int, height: Int, isHorizontal: Boolean = true) :
    Component(x, y, width, height, false) {
    private var isFitParent = false
    var isHorizontal: Boolean = true
        set(value) {
            field = value
            onChange()
        }
    var percent = 0.0
        set(value) {
            field = value.coerceAtMost(100.0)
            onChange()
        }
    var color: PixelColor = PixelColor()
        set(value) {
            field = value
            onChange()
        }

    init {
        color = PixelColor(Color.BLUE)
        this.isHorizontal = isHorizontal
        render()
    }

    override fun render() {
        if (isFitParent && parent != null) {
            parent?.let {
                setSize(it.clientWidth, it.clientHeight)
            }
        } else {
            pixels = arrayOfNulls(width * height)
        }
        if (!isHorizontal) {
            var i = 0
            while (i < pixels.size * percent * 0.01) {
                pixels[i] = color
                i++
            }
        } else {
            var pixelNum: Long = 1
            val width = width
            val height = height
            for (x in 0 until width) {
                if (pixelNum > pixels.size * percent * 0.01) break
                for (y in height - 1 downTo 0) {
                    if (pixelNum > pixels.size * percent * 0.01) break
                    pixels[x + y * width] = color
                    pixelNum++
                }
            }
        }
    }
}